xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/internal/DefaultVelocityManager.java

```diff
@@ -0,0 +1,264 @@
+    {
+        // Set reserved bindings
+
+        // "context" is a reserved binding in JSR223 world
+        this.reservedBindings.add("context");
+
+        // Macros directive
+        this.reservedBindings.add("macro");
+        // Foreach directive
+        this.reservedBindings.add("foreach");
+        // Evaluate directive
+        this.reservedBindings.add("evaluate");
+        // TryCatch directive
+        this.reservedBindings.add("exception");
+        this.reservedBindings.add("try");
+        // Default directive
+        this.reservedBindings.add("define");
+        // The name of the context variable used for the template-level scope
+        this.reservedBindings.add("template");
+    }
+
+    private RuntimeInstance getRuntimeInstance() throws XWikiVelocityException
+    {
+        if (this.runtimeInstance == null) {
+            this.runtimeInstance = initializeRuntimeInstance(this.componentManager, this.velocityConfiguration);
+        }
+
+        return this.runtimeInstance;
+    }
+
+    private synchronized RuntimeInstance initializeRuntimeInstance(ComponentManager componentManager,
+        VelocityConfiguration velocityConfiguration) throws XWikiVelocityException
+    {
+        RuntimeInstance runtime = new RuntimeInstance();
+
+        // Add the Component Manager to allow Velocity extensions to lookup components.
+        runtime.setApplicationAttribute(ComponentManager.class.getName(), componentManager);
+
+        // Set up properties
+        initializeProperties(runtime, velocityConfiguration.getProperties());
+
+        // Set up directives
+        runtime.loadDirective(TryCatchDirective.class.getName());
+
+        try {
+            runtime.init();
+        } catch (Exception e) {
+            throw new XWikiVelocityException("Cannot start the Velocity engine", e);
+        }
+
+        return runtime;
+    }
+
+    /**
+     * @param runtime the Velocity engine against which to initialize Velocity properties
+     * @param configurationProperties the Velocity properties coming from XWiki's configuration
+     */
+    private void initializeProperties(RuntimeInstance runtime, Properties configurationProperties)
+    {
+        // Avoid "unable to find resource 'VM_global_library.vm' in any resource loader." if no
+        // Velocimacro library is defined.
+        runtime.setProperty(RuntimeConstants.VM_LIBRARY, "");
+
+        // Configure Velocity by passing the properties defined in this component's configuration
+        if (configurationProperties != null) {
+            runtime.setProperties(configurationProperties);
+        }
+    }
+
+    /**
+     * @param runtimeInstance the actual Velocity instance to use as reference
+     * @param name the name of the Velocity script
+     * @param source the source of the Velocity script
+     * @return the compiled script
+     */
+    public static VelocityTemplate compile(String name, Reader source, RuntimeInstance runtimeInstance)
```

Comment:
Static method inside a component? that's a bit surprising.

Reply:
> Static method inside a component? that's a bit surprising.

We have many of those, actually. I just needed to put it in a place where I can reuse it and I don't want to expose that as an API.